OPNET-268: Configure-ovs Alternative

cybertron opened the pull request:
There are a number of limitations of the configure-ovs script. This is a proposed design to address them.

cgoncalves commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+
+### Risks and Mitigations
+
+Because this gives the deployer a great deal of control over the bridge
```
On the positive side, should the deployer apply a configuration that renders the node inaccessible, the trio (NetworkManager, nmstate and kubernetes-nmstate) are expected to recover by rolling back to the previous network config state.

cybertron replied:
Good point. I'll mention that as a mitigation to the risk.

cgoncalves commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+
+### Open Questions [optional]
+
+* There has been resistance to using the NMState API in OpenShift in the past
+  because it is not compliant with the OpenShift API Guidelines. There is work
+  underway in NMState to address this, but is that mandatory for this to be
+  implemented?
```
Could you please share some references to both statements?

cybertron replied:
I'll add a link here, but for my own reference it first came up in this thread: https://github.com/openshift/enhancements/pull/1267#discussion_r1013320148

There is an initial PR to smooth integration with OCP here: https://github.com/nmstate/nmstate/pull/2338 It doesn't actually change any of the API yet, but I believe it was a first step in that direction.

cgoncalves commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+  be done via Kubernetes-NMState maybe that's okay? We'll apply them once and
+  never touch them again.
+
+* If changes are made to the bridge configuration on day 2, will OVNKubernetes
+  handle those correctly without a reboot? If not, how do we orchestrate a
+  reboot?
```
OVN-Kubernetes will not handle the replacement of the gateway interface (`enp2s0` in the example above). This is an existing limitation where OVN-Kubernetes installs an OpenFlow rule in br-ex that states the gateway interface is the OVS port ID of `enp2s0` (port 1) and where attaching a new interface, say `enp3s0`, its OVS port ID will be different hence no egress traffic forwarding.

Monitoring br-ex port attachments and recalculating the OpenFlow rule may help. This is outside the scope of this enhancement, though.

cybertron replied:
Ack, thanks. I'm just going to include your comment in the enhancement.

cybertron replied:
I'm wondering if that's because configure-ovs uses different profile names from NMState. It's possible we ended up with multiple profiles doing the same thing.

Maybe I just need to test this. I can deploy my prototype and change the MTU or something like that and see what happens.

jcaamano replied:
At the very least I recall nmstate changing the names of things, like ovs ports and such. Yeah, I am curious about testing this again.

cgoncalves commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+
+### Non-Goals
+
+Complete replacement of configure-ovs. This mechanism is intended for more
```
What are the cases where configure-ovs.sh would still be used?

cybertron replied:
Probably most single-nic deployments where the logic for creating br-ex is simple. I don't anticipate this new mechanism being widely used in AWS, Azure, or GCP, for instance. In fact, I'm not sure how it will work in non-on-prem envs.

cgoncalves replied:
Agreed. Can we add that as an example to this doc?

cybertron replied:
Sure, done.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+### Goals
+
+Once this design is implemented, deployers of OpenShift will be able to
+explicitly configure br-ex to their exact specifications and will be able to
```
It will be mandatory if you're using this new mechanism, but configure-ovs will still exist. I think manual configuration of br-ex is probably too big a hurdle for some deployers so I think we need to keep that easy path. I'm open to being convinced otherwise, but we'll need to discuss that with the OVNK folks.

cybertron replied:
Yep, we're definitely going to want to document what a br-ex config using this mechanism should look like, probably including some more advanced options too.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+          enabled: false
+          dhcp: false
+
+The name field will need to be something that uniquely identifies a given host.
+I'm unsure if name is necessarily the best way to do that, but it is one valid
+option. I reused the networkConfig name from the baremetal feature for
+consistency. Perhaps something different would be preferable to avoid confusion
+though?
```
The issue is that there's no guarantee the name specified in install-config matches the eventual hostname of the node. At least on baremetal, hostnames are controlled by the DHCP server which is outside our control. It's probably fair to say the names need to match if you're going to do this, but I'm open to other options that would reduce the opportunity for misconfiguration too.

cgoncalves replied:
Good point on hostname change because of DHCP.

qinqon replied:
Why is that different from what we have now ? I understand that we are already using the NMState version per node that generated the NM connection files.

cybertron replied:
That's baremetal-specific, we're trying to make this more platform-agnostic. In baremetal we can know the hostnames ahead of time, but I don't know if that's true for all platforms. I suspect there may be some randomness to the hostnames in non-baremetal cases.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+consistency. Perhaps something different would be preferable to avoid confusion
+though?
+
+Note that although the configuration can be provided on a per-host basis, it is
+not mandatory to do so. YAML anchors can be used to duplicate a single
+configuration across multiple nodes. For example:
+
+networking:
+  networkType: OVNKubernetes
+  machineNetwork:
+  - cidr: 192.0.2.0/24
+  [...]
+  hostConfig:
+  - name: master-0
+    networkConfig: &BOND
+      interfaces:
+      - name: br-ex
+        ...etc...
+  - name: master-1
+    networkConfig: *BOND
+  - name: master-2
+    networkConfig: *BOND
+
+This reduces duplication in install-config and the chance of errors due to
+unintended differences in node config.
```
I pre-emptively included this because it was a point of discussion on the baremetal networkConfig design. In that case we agreed to do common config across nodes as a followup, but since it turns out YAML can already handle that for us I thought I'd mention it in case anyone was wondering about that here too.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+This could be implemented almost entirely with machine-configs, with only a simple
+service on the host to apply the provided NMState files. This is the approach
+the [prototype implementation](https://docs.google.com/document/d/1Zp1J2HbVvu-v4mHc9M594JLqt6UwdKoix8xkeSzVL98/edit#heading=h.fba4j9nvp0nl)
+used since per-host configuration was not available any other way. However,
```
- Machine-configs are applied by ignition, the same as this.
- That's a good point. If you scaled out a machine config pool on day 2 you'd have to reboot all of the nodes in that pool because you'd have to add the new node configs to all of them. That's a pretty big limitation.
- True, although in this case I envision the machine-configs writing NMState YAML instead of raw nmconnection files, so you'd still have NMState validating them before it applies them and rolling back if they break things. You couldn't catch syntax errors in install-config though because the installer would just see opaque machine-configs, not their contents.

Actually, the scaleout case is something we need to consider. At install time we can handle mapping things in the installer, but on day 2 we'll need a manual process, much like baremetal has. Unfortunately I don't think we can use kubernetes-nmstate for scaleout either because it won't be running soon enough to apply the bridge config. :-/

cybertron replied:
Note that scaleout works fine with my latest machine-config prototype. You just update the MC with the new node configs and it gets rolled out reboot-less so it will be there when the new nodes start. I discuss how this works below now.

cybertron commented:
/cc @jcaamano @trozet

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+### Workflow Description
+
+When the user is populating install-config they will provide the necessary
+configuration in the networking section (TODO: Is this the right place?).
```
Do you mean a new networkConfig section in install-config? We can't use the existing one for this functionality because we don't have OVS in the ramdisk.

cgoncalves replied:
You say we cannot use the existing one (referring to `networkConfig`, right?), yet the example below with br-ex is in the `networkConfig` block. I'm confused.

cybertron replied:
Yeah, I reused the name but it's in a different section. I don't know if that's more or less confusing than coming up with a new name for a very similar feature.

cybertron commented:
New revision with updates based on the discussion so far. I also left some more thoughts on the variation where this is mostly implemented with machine-configs. I'm actually liking that more and more as I think about it because it uses existing configuration mechanisms so we aren't going to have to invent something completely new.

cybertron commented:
I've pushed an update based on the implementation I did in https://github.com/cybertron/machine-config-operator/tree/configure-ovs-alt

It handles both initial deployment and scaling in a sane fashion and required minimal code changes. The main concern is whether we can use hostname-based mapping of configs on non-baremetal platforms since I think at least some of them add a random identifier into the node names. If we can't use hostnames, does anyone have suggestions on what we _could_ use?

phoracek commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,562 @@
+          enabled: false
+        ipv6:
+          enabled: false
+      - name: br-ex
```
I don't know if it is relevant for this design, but in case the configure-ovs script today also needs to configure ovn-bridge-mappings you may find this useful - we have recently introduced a nice API for it to nmstate: https://bugzilla.redhat.com/show_bug.cgi?id=2209690

With this, you could configure the mapping easily from this CR:

```yaml
hostConfig:
  ovn:
    bridge-mappings:
      - localnet: FOO
        bridge: br-ex
```

cybertron replied:
Yeah, that's definitely a nice fit for this. It would be a pain to introduce an OpenShift-specific API to do that, but with NMState we get it for "free". :-)

qinqon commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,562 @@
+
+#### Variation [optional]
+
+This could be implemented almost entirely with machine-configs, with only a simple
+service on the host to apply the provided NMState files. This is the approach
```
The systemd service is already in place as part of nmstate installation, we just need to put yamls at `/etc/nmstate/`

cybertron replied:
It is, but I wonder if that does what we need. I only want this to apply once, and only the configuration that applies to the specific host. If we implemented this with machine-configs, all of the configs for all of the nodes will be present in the directory. I also only want this to be applied at first boot, and then if changes are needed they should be made by kubernetes-nmstate. We _could_ allow updating via machine-config, but then we throw out all of the benefits that kubernetes-nmstate gives us.

qinqon replied:
Stuff under `/etc/nmstate` apply only once, then nmstate move the file so it's not apply at reboot, also I understand that machine config will dump there just the one for that node.

cybertron replied:
Unfortunately there's no way to tell MCO to only deploy the config specific to the node. We have to write all of the configs to all of the nodes and have logic on the node itself to pick the right one. I've added some discussion of how we might integrate with the NMState service to do this in the Variation section.

qinqon commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,562 @@
+### Risks and Mitigations
+
+Because this gives the deployer a great deal of control over the bridge
+configuration for OVNK, they will have the ability to configure it incorrectly,
+which may cause problems either on initial deployment or at any point after.
+We will want to ensure that must-gather is collecting enough information about
+the bridge configuration for us to determine whether it is compliant with our
+guidelines, and we must have clear instructions on what is required for br-ex.
```
To reduce this risk we should document nmstate examples to create br-ex manually for differente scenarios (single-nic, multi-nic, bond + bridge + vlans, etc...)

We have to be super careful with MTUs though.

cybertron replied:
Yep, probably similar to what we have for kubernetes-nmstate today. We'll include examples of some common configs.

qinqon commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,562 @@
+This is a large departure from that philosophy, but that is unavoidable.
+Certain aspects of networking (notably static IPs) are by nature host-specific
+and cannot be handled any other way.
+
+Using a machine-config to do initial deployment and a
+NodeNetworkConfigurationPolicy to make day 2 changes will require a manual sync
+of the NMState configuration. Unfortunately this is unavoidable because the
+Kubernetes-NMState operator is not part of the core payload so it (and its
+CRDs) cannot be used for initial deployment.
+
```
But we can change k-nmstate operator to read stuff under /etc/nmstate and convert it to NNCPs.

cybertron replied:
That's an interesting idea. We could do that with the baremetal feature because only the nmconnection files made it to the host, but in this case we'll have the full NMState config available. It would be very interesting to add a feature to discover the existing config for the operator.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,562 @@
+At a high level, we will provide an interface for the deployer to specify
+per-host interface configuration. Configuration for an OpenVSwitch bridge
+named br-ex will be mandatory. The format for the configuration will be
+NMState YAML, which will be provided to each node via Ignition. Around the
+same point in the boot process that configure-ovs would run, we will run
+NMState to apply the provided configuration. This will only be done on first
+boot because subsequent changes will be made using the Kubernetes-NMState
+operator.
```
Yep, will do.

Right now it would need to be done manually. There isn't really a way to convert it at install time because kubernetes-nmstate won't be installed yet. I suppose there may be some way to auto-create some manifests that could be applied later though.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,562 @@
+  [...]
+  hostConfig:
+  - name: master-0
+    networkConfig:
```
I suspect we would use the results of nodeip-configuration to generate the policy. I'm not sure exactly how that would work, but it's definitely something I'd like to look into. It's not my primary goal right now because I want to handle the advanced cases where configure-ovs struggles, but I do see nmpolicy as a potential way to completely replace configure-ovs.

On a more general note, a per-cluster config (maybe with the possibilty of overriding on a per-node basis) may be a solution to platforms where we don't have a way to map per-node configs ahead of time (like if the hostname is not predictable). I don't know how important this feature will be in those cases, but I have heard some discussion of needing to override the MTU at deploy time, so this might provide a solution to that. I think with my current machine-config prototype it would be pretty easy to extend to include a per-cluster config so maybe I'll give that a try and update the design if it works well.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,562 @@
+  machineNetwork:
+  - cidr: 192.0.2.0/24
+  [...]
+  hostConfig:
+  - name: master-0
+    networkConfig: &BOND
+      interfaces:
+      - name: br-ex
+        ...etc...
+  - name: master-1
+    networkConfig: *BOND
+  - name: master-2
+    networkConfig: *BOND
```
Agreed. I'm seriously tempted to drop the install-config part of this entirely for the initial implementation since we can do it all with machine-configs, but even if we do that I think a cluster-wide config makes a lot of sense.

I think day 2 would be handled similar to how it is today, which is the user would need to manually copy their config into an NNCP. I want to eliminate that step, but I think we'll need to get KNMState into the core payload first.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,498 @@
+This reduces duplication in install-config and the chance of errors due to
+unintended differences in node config.
+
+When this mechanism is in use, a flag will be set to indicate to configure-ovs
```
Oops, did I not reply to this before?

The answer is yes. configure-ovs will still run in the default case, but if you've provided custom br-ex config it will prevent configure-ovs from running.

cybertron reviewed (commented):
Thanks for all the comments! I will update this and the prototype accordingly.

cybertron commented:
I've updated the document to reflect the fact that I would like to implement the machine-config method as an MVP, and then work on improved integration and more features as followups. I think that will let us get a useful feature into the hands of users more quickly, and we can address any usability shortcomings based on real-world feedback.

jcaamano commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,568 @@
+* As an OpenShift developer, I want a mechanism for bridge configuration that
+  is well-tested and less likely to break than a Bash script.
+
+* As an OpenShift developer, I want to simplify the configure-ovs.sh script
+  by providing a better way to do advanced bridge configuration.
```
I guess we cannot remove any currently supported use case from configure-ovs if we want to keep backward compatibility. So I guess this is more about an new way of supporting new advanced use cases without complicating configure-ovs more than it already is.

jcaamano commented on enhancements/network/configure-ovs-alternative.md:
````diff
@@ -0,0 +1,568 @@
+        path: /etc/nmstate/cluster.yaml
+```
+
+The only difference is there will be a single file applied to each node of the
````
I read above that the nmstate service moves the files away after they are applied. So we wouldn't need to worry about this if we used that.

cybertron replied:
I was looking at that as an implementation detail. My current implementation just writes a file after we process the configuration once (which we'll still need to do even using the default NMState service because we need to make sure it only applies the configuration for the current node).

The file is convenient though because configure-ovs can look for it to determine whether we're using this new method.

jcaamano replied:
Understood. I guess I was being overzealous in crosschecking if that would be something a user might inadvertently revert but I guess there is no need to worry about that.

jcaamano replied:
Also occurred to me that if we use the nmstate service and that moves the files around, MCO might complain.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,568 @@
+custom service that will select only the host-specific NMState file and
+apply it. This is what my [current prototype](https://docs.google.com/document/d/1Zp1J2HbVvu-v4mHc9M594JLqt6UwdKoix8xkeSzVL98/edit#heading=h.fba4j9nvp0nl) does.
+
+The other is to use the NMState service already included in the OS image to
+apply the configurations. One issue with this is that it will apply everything
+in the /etc/nmstate directory rather than just the host-specific file.
+To get around that, we could still implement a custom service, but instead of
+applying the configurations, it would only be responsible for copying the
+correct file to /etc/nmstate so the regular service can apply it. This has
+the benefit of not reinventing the NMState service and still gives us full
+control over what gets applied to the host, and also leaves us a way to signal
+to ovs-configuration that we've already handled the bridge setup.
```
We generally don't support mixing network configuration mechanisms. Still, it would probably be safest to verify there is no conflict with files already in /etc/nmstate to avoid overwriting a file we shouldn't. In that case we should probably just fail the service.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,568 @@
+  Reply from @cgoncalves:
+  > OVN-Kubernetes will not handle the replacement of the gateway interface (enp2s0 in the example above). This is an existing limitation where OVN-Kubernetes installs an OpenFlow rule in br-ex that states the gateway interface is the OVS port ID of enp2s0 (port 1) and where attaching a new interface, say enp3s0, its OVS port ID will be different hence no egress traffic forwarding.
+
+  That doesn't have to block this feature being implemented, but it's something
+  we should look into as a followup with the OVNK team.
```
In the MTU case, is configure-ovs responsible for changing the MTU or does it just pick up the changed MTU of the underlying interface? I was thinking that you had to manually modify the interface as part of the process, which you would just do with Kubernetes-NMState here.

It would be good if we can come up with a list of day 2 features that might require bridge changes so we know what use cases to target.

jcaamano replied:
It just picks up the changed MTU of the underlying interface that a user is responsible for configuring correctly.

> which you would just do with Kubernetes-NMState here

If you mean changing br-ex and children interfaces, then we need to make sure this works with Kubernetes-NMState. That's what I meant that needs to be considered within the scope of the enhancement.

I am aware of any other features that depend directly on changing network configuration on day 2 (perhaps this could also help when enabling dualstack) but I do know some other features depend on the MTU migration itself, like enabling ipsec or sdn<->ovn migration.

jcaamano commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,568 @@
+named br-ex will be mandatory. The format for the configuration will be
+NMState YAML, which will be provided to each node via Ignition. Around the
+same point in the boot process that configure-ovs would run, we will run
+NMState to apply the provided configuration. This will only be done on first
```
So let's say that you copy the entire day 1 config into an NNCP and make changes to that NNCP on day 2. Would that result in nmstate operator writing back updated config to /etc/NetwrokManager/system-connections? Or would the node, upon reboot, use the original day 1 configuration in /etc/NetwrokManager/system-connections and then once the nmstate operator is running again would hte configuraiton be reconciled?

cybertron commented:
Unfortunately it looks like applying an NNCP to br-ex, even if there are no changes from the day 1 configuration, still breaks connectivity. That's a pretty significant problem for this design, which means one of two things will need to happen:

1. We change the scope so that this only applies to day1 and you are still not allowed to change br-ex using the operator.
2. We find a way to make updates with the operator work. Unless there's some quick fix (which I doubt), this option likely means missing 4.15.

jcaamano commented:
> Unfortunately it looks like applying an NNCP to br-ex, even if there are no changes from the day 1 configuration, still breaks connectivity. That's a pretty significant problem for this design, which means one of two things will need to happen:
> 
> 1. We change the scope so that this only applies to day1 and you are still not allowed to change br-ex using the operator.
> 2. We find a way to make updates with the operator work. Unless there's some quick fix (which I doubt), this option likely means missing 4.15.

I was just going to update here that I tried this today and I did not get very far. :/ I would like to touch base with @qinqon on this next week.

> 1. We change the scope so that this only applies to day1 and you are still not allowed to change br-ex using the operator.

You would still be allowed to make day-2 configuration changes on the resulting NM configuration that should live in /etc/NetworkManager/system-connections right? Which is probably not much more difficult than handling the source nmstate manifests.

qinqon commented:
> Unfortunately it looks like applying an NNCP to br-ex, even if there are no changes from the day 1 configuration, still breaks connectivity. That's a pretty significant problem for this design, which means one of two things will need to happen:
> 
> 1. We change the scope so that this only applies to day1 and you are still not allowed to change br-ex using the operator.
> 2. We find a way to make updates with the operator work. Unless there's some quick fix (which I doubt), this option likely means missing 4.15.

@cybertron can you dump here the original nmstate from br-ex and what change did you try ? I hink nmstate team has fix some issues related to ovs/ovn like https://docs.rs/nmstate/latest/nmstate/struct.OvsBridgeConfig.html#structfield.allow_extra_patch_ports

cybertron commented:
Here's the NMState YAML I used. This is the same both on day 1 and on day 2 in the NNCP.

```interfaces:
- name: enp2s0
  type: ethernet
  state: up
  ipv4:
    enabled: false
  ipv6:
    enabled: false
- name: br-ex
  type: ovs-bridge
  state: up
  copy-mac-from: enp2s0
  ipv4:
    enabled: false
    dhcp: false
  ipv6:
    enabled: false
    dhcp: false
  bridge:
    port:
    - name: enp2s0
    - name: br-ex
- name: br-ex
  type: ovs-interface
  state: up
  ipv4:
    enabled: true
    address:
    - ip: "192.168.111.110"
      prefix-length: 24
  ipv6:
    enabled: false
    dhcp: false
dns-resolver:
  config:
    server:
    - 192.168.111.1
routes:
  config:
  - destination: 0.0.0.0/0
    next-hop-address: 192.168.111.1
    next-hop-interface: br-ex```

jcaamano commented:
We tested day-2 configuration changes with Kubernetes NMstate and got them more-less working with this fix:
https://github.com/openshift/machine-config-operator/pull/3998
More potential fixes might be needed in Kubernetes NMstate for smoother operation.

There is still the open question of how to quantify the possible disruption of making day-2 configuration changes with Kubernetes NMstate on the default network when compared to the procedure that could be done today which implies a rolling reboot.

I think that the only day-2 change documented today is the MTU migration procedure:
https://docs.openshift.com/container-platform/4.13/networking/changing-cluster-network-mtu.html

So we could try it out. Overall I think we need to cover that procedure for users using the mechanism described in this enhancement.

dhellmann commented:
(automated message) This pull request is closed with lifecycle/rotten. The associated Jira ticket, OPNET-265, has status "In Progress". Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

dhellmann commented:
(automated message) This pull request is closed with lifecycle/rotten. The associated Jira ticket, OPNET-265, has status "In Progress". Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

dhellmann commented:
(automated message) This pull request is closed with lifecycle/rotten. The associated Jira ticket, OPNET-265, has status "In Progress". Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

cybertron commented:
/reopen

dhellmann commented:
#1555 is changing the enhancement template in a way that will cause the header check in the linter job to fail for existing PRs. If this PR is merged within the development period for 4.16 you may override the linter if the only failures are caused by issues with the headers (please make sure the markdown formatting is correct). If this PR is not merged before 4.16 development closes, please update the enhancement to conform to the new template.

cybertron commented:
/remove-lifecycle rotten

I do still need to update this to reflect the new template though.

cybertron commented:
Rebased and updated to pass lint. I filled out the TBDs for reviewers and approvers, but I'm happy to change those if someone else would be more appropriate.

cgoncalves commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,407 @@
+Certain aspects of networking (notably static IPs) are by nature host-specific
+and cannot be handled any other way.
+
+Using a machine-config to do initial deployment and a
```
How can existing nodes move off configure-ovs to this new approach?

Do cluster admins have to create the `10-br-ex-worker` machine-config (will reboot potentially all nodes) with the desired new network configuration (and be able to further update it on day 2 via k-nmstate)?

Apologies if it's stated somewhere in the doc. In any case, we would want to document such a procedure on the OCP documentation page.

cybertron replied:
I've added a discussion of how I expect migration to work and opened a story on the epic to track verification that this works as intended: https://issues.redhat.com/browse/OPNET-508

jcaamano commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,407 @@
+
+When the new service runs, it will copy the appropriate YAML file (based on
+the hostname of the node) to /etc/nmstate so it will be applied by the normal
+system NMState service. If there is already a file in /etc/nmstate with the
+same name, the service will fail in order to avoid overwriting important
```
I wonder if we should check just for a file of the same name or just any file:

- Do we consider the the method described here as the only supported mechanism to deploy nmstate config on the host?
- What if there was some configuration already there from a system on a previous version that was upgraded to this version and user wants to start using this mechanism?

What I mean here is that if there are other files we can't really be sure if there's any conflict. So I feel this should be a completely exlusive mechanism and warned as so in the docs.

cybertron replied:
Agreed, and this is already our recommendation for kubernetes-nmstate because it is problematic to have multiple separate configurations attempting to configure the same host. I will add to this section to be explicit that when using the new mechanism for configuration we expect that to be the only network configuration used.

jcaamano commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,407 @@
+When this mechanism is in use, a flag will be set to indicate to configure-ovs
+that it should not attempt to configure br-ex.
+
+Day 2 changes will be handled by Kubernetes-NMState. For the initial
```
Two questions here:

- Do we prevent in any way processing day-2 changes on the MachineConfigs described in this procedure? If not, would that be an alternative mechanism to deploy day-2 changes if user is willing to put up with a reboot? I guess the answer is yes, to allow for the scale out scenario.
- Do we prevent day-2 with Kubernetes-NMState to deployments that are still relying on configure-ovs

cybertron replied:
- A reboot won't happen on day 2 changes to these files because we are adding the /etc/nmstate/openshift path to the reboot-less update list in MCO. We won't otherwise prevent changes to these files, but changes to existing nodes will have no effect because these configs are only applied on first boot.
- We have a warning in the kubernetes-nmstate docs that modification of br-ex and the underlying interface is not allowed. We'll need to modify that to allow changes if this new feature is used instead of configure ovs.

jcaamano commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,407 @@
+changes:
+
+- DNS updates
+- MTU updates
```
There is a very little nit handling on configure-ovs tied to the MTU migration procedure:

https://github.com/openshift/machine-config-operator/blob/204767253e30608b5b7fd70ad1ace02ba1d64b46/templates/common/_base/files/configure-ovs-network.yaml#L880-L884

cybertron replied:
Thanks for the heads up. I will make sure this is handled in the implementation patch.

jcaamano commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,407 @@
+
+#### Standalone Clusters
+
+This is primarily directed at standalone clusters, specifically baremetal and
```
This might be confusing. Is this supported in any cluster or just non-cloud?

cybertron replied:
I will rewrite this to be clearer.

cybertron commented on enhancements/network/configure-ovs-alternative.md:
```diff
@@ -0,0 +1,407 @@
+
+### Tech Preview -> GA
+
+Our primary concern for GA at this time is error-handling. In particular,
```
Done.

cybertron commented:
I think this should address all of the comments so far. Let me know what you think.

I'll also be updating the implementation patch today based on what we've discussed here and in the sync meeting.

cgoncalves commented:
LGTM. Thank you!

jcaamano commented:
/lgtm

qinqon commented:
/lgtm

yuqi-zhang reviewed (commented):
Functionally I think the proposal makes sense. Most of the MCO's concerns are around having a proper API workflow around this, which as I understand is in long-term design considerations. Just leaving some notes based on our discussions on slack so we keep those in mind

knobunc commented:
/approve
